### PR TITLE
[RFC] Introduce Response Content Matcher and Response Loader

### PR DESCRIPTION
**Pull request by Arminek**
I decided to extract some logic from api test case for future integration.

### CODE REVIEW

File: src/ApiTestCase.php [hunk 1]
```diff
@@ -96,6 +108,8 @@ public function setUp()
     {
         parent::setUp();
         $this->fixtureLoader = new Loader();
+        $this->responseContentMatcher = new ResponseContentMatcher(new DefaultMatcher());
```
Comment (lchrusciel) [inline src/ApiTestCase.php]: Won't it be better to extract it to separate set up method? `setUpMatchet` for instance? This way we won't be forced to copy-paste whole setUp, and instead we will be able to override only direct setUp. WDYT?

File: src/ApiTestCase.php [hunk 2]
```diff
@@ -376,6 +382,26 @@ private function assertSourceExists($source)
     }
 
     /**
+     * @throws \RuntimeException
+     */
+    private function assertResponseLoaderIsAvailable()
+    {
+        if (null === $this->responseLoader) {
+            throw new \RuntimeException('Something went wrong and response loader did not set properly, if you have overridden setUp method check it!');
```
Comment (lchrusciel) [inline src/ApiTestCase.php]: `response loader has not been set properly. If you have overridden setUp method, check it!`?

File: src/Services/MatcherInterface.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * @author Arkadiusz Krakowiak <arkadiusz.krakowiak@lakion.com>
+ */
+interface MatcherInterface
+{
+    /**
+     * @param $value
+     * @param $pattern
+     *
+     * @return mixed
```
Comment (lchrusciel) [inline src/Services/MatcherInterface.php]: Shouldn't match return boolean?

File: src/Services/ResponseContentMatcher.php [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+ */
+class ResponseContentMatcher implements ResponseContentMatcherInterface
+{
+    /**
+     * @var MatcherInterface
+     */
+    private $matcher;
+
+    /**
+     * @var null|string
```
Comment (lchrusciel) [inline src/Services/ResponseContentMatcher.php]: just `string`

File: src/ApiTestCase.php (hunk 1 above)
Comment (Arminek) [inline src/ApiTestCase.php]: :+1:

**Comment by Arminek** [comment]
@lchrusciel what about namespace ? I was not sure where i should place those services.

File: src/Services/ResponseLoader.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Lakion\ApiTestCase\Services;
```
Comment (lchrusciel) [inline src/Services/ResponseLoader.php]: `Lakion\ApiTestCase\Loader`

File: src/Services/ResponseContentMatcherInterface.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Lakion\ApiTestCase\Services;
```
Comment (lchrusciel) [inline src/Services/ResponseContentMatcherInterface.php]: `Lakion\ApiTestCase\Matcher`

**Comment by lchrusciel** [comment]
@Arminek I add comments with my proposal. Haven't seen it before :dash:

**Comment by Arminek** [comment]
:+1:

File: src/Loader/ResponseLoader.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    /**
+     * @param string $source
+     *
+     * @throws \RuntimeException
+     */
+    private function assertSourceIsNotFolder($source)
+    {
+        if (true === is_dir($source)) {
```
Comment (pamil) [inline src/Loader/ResponseLoader.php]: What if it is a device instead? :)

Comment (lchrusciel) [inline src/Loader/ResponseLoader.php]: `is_file`?

Comment (Arminek) [inline src/Loader/ResponseLoader.php]: What do you mean? I think `->assertContentIsProperLoaded()` should catch this :)

Comment (pamil) [inline src/Loader/ResponseLoader.php]: But `file_get_contents` would throw an error anyway. We can isolate it inside a private method:
```php
private function getContents($file)
{
  $contents = @file_get_contents($file);
  if (false === $contents) {
    throw new \Exception();
  }

  return $contents;
}
```

File: src/Matcher/ResponseContentMatcher.php [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+     */
+    public function __construct(MatcherInterface $matcher)
+    {
+        $this->matcher = $matcher;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function matchResponse($actualResponse, $expectedResponse)
```
Comment (pamil) [inline src/Matcher/ResponseContentMatcher.php]: That method is quite counterintuitive. Let's say I want to match a response, I call it with wrong response and nothing happens. It results in a stateful code that we must know very well before using.

Why don't we throw a dedicated exception if response does not match the expected one?

File: src/Matcher/DefaultMatcher.php [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+namespace Lakion\ApiTestCase\Matcher;
+
+use Coduo\PHPMatcher\Factory\SimpleFactory;
+use Coduo\PHPMatcher\Matcher;
+
+/**
+ * @author Arkadiusz Krakowiak <arkadiusz.krakowiak@lakion.com>
+ */
+class DefaultMatcher implements MatcherInterface
```
Comment (pamil) [inline src/Matcher/DefaultMatcher.php]: It's not the `DefaultMatcher`, it's rather `CoduoMatcher` or `CoduePHPMatcher`?

File: src/Matcher/DefaultMatcher.php [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+{
+    /**
+     * @var Matcher
+     */
+    private $matcher;
+
+    public function __construct()
+    {
+        $factory = new SimpleFactory();
+        $this->matcher = $factory->createMatcher();
```
Comment (pamil) [inline src/Matcher/DefaultMatcher.php]: `$this->matcher = (new SimpleFactory())->createMatcher();`

File: src/Matcher/DefaultMatcher.php [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function match($value, $pattern)
```
Comment (pamil) [inline src/Matcher/DefaultMatcher.php]: Same comment as for `ResponseContentMatcher::matchRepsonse`.

File: src/Matcher/ResponseContentMatcher.php (hunk 8 above)
Comment (lchrusciel) [inline src/Matcher/ResponseContentMatcher.php]: anyway `match` would be enough

Comment (pamil) [inline src/Matcher/ResponseContentMatcher.php]: @lchrusciel :+1:, then we can merge `MatcherInterface` and `ResponseContentMatcherInterface` into one, as they do the same job.

Comment (Arminek) [inline src/Matcher/ResponseContentMatcher.php]: :+1:

File: src/Matcher/DefaultMatcher.php (hunk 9 above)
Comment (Arminek) [inline src/Matcher/DefaultMatcher.php]: :+1:

**Comment by lchrusciel** [comment]
@Arminek are you willing to rebase and apply above comments?

**Pull request closed without merging**